service/src/main/java/bio/terra/workspace/service/resource/controlled/flight/clone/bucket/CopyGcsBucketDataStep.java

```diff
@@ -0,0 +1,283 @@
+    FlightUtils.setResponse(flightContext, apiBucketResult, HttpStatus.OK);
+    return StepResult.getStepResultSuccess();
+  }
+
+  private void setSourceBucketRoles(String sourceBucketName, String transferServiceSAEmail) {
+    setBucketRoles(
+        sourceBucket.getWorkspaceId(),
+        sourceBucketName,
+        transferServiceSAEmail,
+        ImmutableSet.of("roles/storage.objectViewer", "roles/storage.legacyBucketReader"));
+  }
+
+  private void setDestinationBucketRoles(
+      UUID destinationWorkspaceId, String bucketName, String transferServiceSAEmail) {
+    setBucketRoles(
+        destinationWorkspaceId,
+        bucketName,
+        transferServiceSAEmail,
+        ImmutableSet.of("roles/storage.legacyBucketWriter"));
+  }
+
+  private void setBucketRoles(
+      UUID workspaceId, String bucketName, String transferServiceSAEmail, Set<String> roles) {
+    final String projectId = workspaceService.getRequiredGcpProject(workspaceId);
+    final StorageCow storageCow = crlService.createStorageCow(projectId);
+    final Identity saIdentity = Identity.serviceAccount(transferServiceSAEmail);
+    final Policy.Builder policyBuilder = storageCow.getIamPolicy(bucketName).toBuilder();
+    roles.forEach(s -> policyBuilder.addIdentity(Role.of(s), saIdentity));
+    storageCow.setIamPolicy(bucketName, policyBuilder.build());
+  }
+
+  // First, we poll the transfer jobs endpoint until an operation has started so that we can get
+  // its server-generated name.
+  private String getLatestOperationName(
+      Storagetransfer storageTransferService, String transferJobName, String projectId)
+      throws InterruptedException, IOException {
+    int attempts = 0;
+    String operationName;
+    do {
+      TimeUnit.MILLISECONDS.sleep(JOBS_POLL_INTERVAL.toMillis());
+      final TransferJob getResponse =
+          storageTransferService.transferJobs().get(transferJobName, projectId).execute();
+      operationName = getResponse.getLatestOperationName();
+      attempts++;
+      if (attempts > MAX_ATTEMPTS) {
+        throw new RuntimeException("Exceeded max attempts to get transfer operation name");
+      }
+    } while (operationName == null);
+    logger.info("Latest transfer operation name is {}", operationName);
+    return operationName;
+  }
+
+  private TransferSpec createBulkTransferSpec(
+      String sourceBucketName, String destinationBucketName) {
+    return new TransferSpec()
+        .setGcsDataSource(new GcsData().setBucketName(sourceBucketName))
+        .setGcsDataSink(new GcsData().setBucketName(destinationBucketName))
+        .setTransferOptions(
+            new TransferOptions()
+                .setDeleteObjectsFromSourceAfterTransfer(false)
+                .setOverwriteObjectsAlreadyExistingInSink(false));
+  }
+
+  @Override
+  public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+    return StepResult.getStepResultSuccess();
```

Comment:
Right now the undo in the previous step deletes the destination bucket. I'm not sure if we want to keep that, because of the possibility of very expensive transfers failing at the eleventh hour.

Reply:
I suspect the client can't reuse their proposed destination bucket name if we delete it, and right now I don't allow you to start a transfer into an existing bucket from the API. So if they have to retry client side, they're looking at a whole new destination bucket, and a fresh transfer of everything (and probably an orphaned destination bucket if it's partially full (the delete in the undo will fail) if we're not careful).

Reply:
The undo code in the bucket definition step should handle files in the bucket when deleting the whole controlled resource.